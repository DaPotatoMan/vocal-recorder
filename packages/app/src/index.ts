import type { useAudioProcessor } from './processor'

export function createRecorder(config: Recorder.Config) {
  const { State } = Recorder

  let state = State.inactive
  let processor: Recorder.Processor | undefined

  async function init() {
    if (state === State.initialized || state === State.recording)
      return

    const { useAudioProcessor } = config.legacy
      ? await import('./processor/legacy')
      : await import('./processor')

    processor = await useAudioProcessor(config)
    state = State.initialized
  }

  function start() {
    if (state === State.recording) throw new Error('Recording is already running!')
    if (!processor) throw new Error('Recorder not initialized!')

    processor.start()
    state = State.recording
  }

  async function stop() {
    if (!processor) throw new Error('Recorder not initialized!')

    state = State.inactive
    return processor.stop().finally(dispose)
  }

  async function dispose() {
    if (state !== State.inactive)
      await processor?.dispose()

    state = State.inactive
    processor = undefined
  }

  return {
    init,
    start,
    stop,
    dispose,
    get state() { return state },
    get processor() { return processor }
  }
}

export class Recorder {
  static create(config: Recorder.Config) {
    return createRecorder(config)
  }
}

export namespace Recorder {
  export interface Config {
    /** When `true` the recorder will use native MediaRecorder and `will not encode to mp3` */
    legacy?: boolean

    stream?: MediaTrackConstraints
    wasmPath?: string
    workletPath: string
  }

  export enum State {
    /** Recorder is disposed or not initialized */
    inactive,
    initialized,
    recording
  }

  export type Processor = Awaited<ReturnType<typeof useAudioProcessor>>
  export type Instance = ReturnType<typeof Recorder.create>

  export interface Result {
    blob: Blob
    duration: number
  }
}
