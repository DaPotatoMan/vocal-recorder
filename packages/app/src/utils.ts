export function getAudioCtx(options?: AudioContextOptions) {
  return new (window.AudioContext || window.webkitAudioContext)(options)
}

export function getStream(options?: MediaTrackConstraints) {
  const ctx: any = navigator.mediaDevices || navigator || {}
  const getMedia: MediaDevices['getUserMedia'] = (ctx.getUserMedia || ctx.getUserMedia
    || ctx.webkitGetUserMedia || ctx.mozGetUserMedia || ctx.msGetUserMedia)

  if (!getMedia) throw new Error('getUserMedia is not supported in this device.')

  return getMedia.call(ctx, { audio: options || true, video: false })
}

export function disposeStream(stream?: MediaStream) {
  if (!stream) return false

  stream.getTracks().forEach((track) => {
    track.stop()
    stream.removeTrack(track)
  })

  return stream.active
}
