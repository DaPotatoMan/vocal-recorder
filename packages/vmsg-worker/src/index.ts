import type { WorkerGetType } from './types'
import AudioWorker from './worker?worker'

export function useWorker(options: any) {
  let promise: {
    resolve: (blob: Blob) => void
    reject: (reason?: any) => void
  }

  const worker = new AudioWorker()
  const post = (type: WorkerGetType, data?: any) => worker.postMessage({ type, data })

  const methods = {
    start: (sampleRate: number) => post('start', sampleRate),
    encode: (data: Float32Array) => post('data', data),
    stop() {
      return new Promise<Blob>((resolve, reject) => {
        promise = { resolve, reject }
        post('stop')
      })
    },

    dispose: () => worker.terminate()
  }

  return new Promise<typeof methods>((resolve, reject) => {
    worker.onmessage = (e) => {
      const { data, type } = e.data
      const state: string = type

      if (state === 'init') return resolve(methods)
      if (state === 'stop') {
        methods.dispose()
        return promise.resolve(data)
      }

      if (state.endsWith('error')) {
        methods.dispose()

        console.error('Worker error:', data)
        const error = new Error(data)

        promise?.reject(error)
        reject(error)
      }
    }

    post('init', options)
  })
}
