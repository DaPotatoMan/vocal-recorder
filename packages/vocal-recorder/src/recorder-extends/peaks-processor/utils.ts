export function getAveragePeaks(peaks: number[], samples = 64) {
  const blockSize = Math.floor(peaks.length / samples) // the number of samples in each subdivision
  const filteredData = []

  for (let i = 0; i < samples; i++) {
    const blockStart = blockSize * i // the location of the first sample in the block
    let sum = 0
    for (let j = 0; j < blockSize; j++)
      sum = sum + Math.abs(peaks[blockStart + j]) // find the sum of all the samples in the block

    filteredData.push(sum / blockSize) // divide the sum by the block size to get the average
  }

  // Normalizes the audio data to make a cleaner illustration
  const maxPeak = Math.max(...filteredData)

  // Return raw data if fully silent
  if (maxPeak <= 0.0001)
    return filteredData

  const multiplier = maxPeak ** -1
  return filteredData.map(n => n * multiplier * 100)
}
