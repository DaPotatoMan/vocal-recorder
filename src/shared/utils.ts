import mitt from 'mitt'

export class DeferredPromise<T> extends Promise<T> {
  resolve: (value: T | PromiseLike<T>) => void
  reject: (reason: any) => void

  initialCallStack: Error['stack']
  resolved = false

  constructor(executor: ConstructorParameters<typeof Promise<T>>[0] = () => { }) {
    let resolver: (value: T | PromiseLike<T>) => void
    let rejector: (reason: T | Error) => void

    super((resolve, reject) => {
      resolver = resolve
      rejector = reject
      return executor(resolve, reject) // Promise magic: this line is unexplicably essential
    })

    this.resolve = (...args) => {
      this.resolved = true
      return resolver(...args)
    }

    this.reject = rejector!

    // store call stack for location where instance is created
    this.initialCallStack = new Error(' ').stack?.split('\n').slice(2).join('\n')
  }

  /** @throws error with amended call stack */
  rejectWithError(error: Error) {
    error.stack = [error.stack?.split('\n')[0], this.initialCallStack].join('\n')
    this.reject(error)
  }
}

export function getWindow() {
  // eslint-disable-next-line no-restricted-globals
  return self
}

export function useAsyncQueue() {
  let promise: Promise<unknown> = Promise.resolve()

  return {
    get promise() { return promise },

    run<T>(task: () => T | Promise<T>) {
      return promise = promise.then(task)
    }
  }
}

/**
 * Wrapper for mitt package
 * @private
 */
export function useEvents<T extends Record<any, any>>() {
  return mitt<T>()
}
