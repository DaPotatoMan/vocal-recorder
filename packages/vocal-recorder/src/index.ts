import type { AudioBlob } from './recorder-extends/utils/audio-blob'
import { StreamRecorder } from './recorder-extends/stream'
import { getStream } from './recorder-extends/utils/media'

export type { AudioBlob } from './recorder-extends/utils/audio-blob'
export * from './recorder-extends/peaks-processor/utils'
export { RealtimePeaksProcessor } from './recorder-extends/peaks-processor'

export class Recorder {
  instance!: StreamRecorder

  constructor(public config: Partial<Recorder.Config> = {}) {}

  async init(stream?: MediaStream) {
    this.instance = new StreamRecorder(stream || await getStream(this.config.constraints))
    this.instance.on('stop', () => this.dispose())
  }

  async restart() {
    if (this.instance)
      await this.stop()

    return this.start()
  }

  /** Start/Resume recorder */
  async start() {
    if (!this.instance)
      await this.init()

    if (this.instance.state === 'recording')
      throw new Error('Recording is already running!')

    // Resume if paused
    if (this.instance.state === 'paused')
      this.instance.resume()

    this.instance.start()
  }

  stop() {
    return this.instance.stop()
  }

  resume() {
    return this.instance.resume()
  }

  pause() {
    return this.instance.pause()
  }

  dispose() {
    this.instance = undefined as any
  }

  // Events
  on(...args: Parameters<StreamRecorder['on']>) {
    return this.instance.on(...args)
  }

  off(...args: Parameters<StreamRecorder['off']>) {
    return this.instance.off(...args)
  }

  emit(...args: Parameters<StreamRecorder['emit']>) {
    return this.instance.emit(...args)
  }
}

export namespace Recorder {
  export interface Config {
    constraints?: MediaTrackConstraints
  }

  export type Result = AudioBlob
}
