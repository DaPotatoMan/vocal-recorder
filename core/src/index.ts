import { useProcessor } from './processor'
import { AudioBlob } from './processor/factories'
import { type EventBus, Logger, StreamUtil, getAudioContext, useEvents } from './shared'

function getContextNodes(emitter: EventBus) {
  function build() {
    const context = getAudioContext({ latencyHint: 'playback' })
    const desination = context.createMediaStreamDestination()
    const gainNode = context.createGain()
    const processor = useProcessor(emitter, context, gainNode, desination)

    context.suspend()
    gainNode.connect(desination)

    async function dispose() {
      if (context.state !== 'closed') await context.close()
      desination.disconnect()
      gainNode.disconnect()
    }

    return {
      context, desination, gainNode, processor, dispose
    }
  }

  const nodes = build()

  function rebuildNodes(whenClosed = true) {
    if (whenClosed && nodes.context.state !== 'closed')
      return

    Logger.log('getContextNodes() -> rebuilding nodes')
    nodes.dispose() // ! Dispose current nodes
    return Object.assign(nodes, build())
  }

  return { nodes, rebuildNodes }
}

function useStreamSwitcher(nodes: ReturnType<typeof getContextNodes>['nodes']) {
  let node: MediaStreamAudioSourceNode | null = null

  function dispose() {
    if (!node) return

    node.disconnect()
    StreamUtil.dispose(node.mediaStream)
    node = null
  }

  function use(stream: MediaStream) {
    dispose()

    node = nodes.context.createMediaStreamSource(stream)
    node.connect(nodes.gainNode)
  }

  return {
    use,
    dispose,

    get isValid() {
      return !!node && StreamUtil.isValid(node.mediaStream)
    }
  }
}

function getContextState(context: AudioContext) {
  const state = context.state.toString() as AudioContextState & {}

  return {
    current: state,
    isRecording: state === 'running',
    isPaused: state === 'suspended' && context.currentTime > 0,
    isStopped: state === 'closed'
  }
}

export function useRecorder() {
  const emitter = useEvents()
  const { nodes, rebuildNodes: verify } = getContextNodes(emitter)
  const streamNode = useStreamSwitcher(nodes)

  emitter.on('*', key => Logger.log('state changed ->', key))

  async function init(stream?: MediaStream) {
    verify()

    await nodes.processor.isReady

    stream ||= await StreamUtil.get()
    switchStream(stream)
    emitter.emit('init')
  }

  async function start() {
    verify()
    if (!streamNode.isValid) await init()

    await nodes.context.resume()
    emitter.emit('start')
  }

  async function resume() {
    verify()

    await nodes.context.resume()
    emitter.emit('resume')
  }

  async function pause() {
    await nodes.context.suspend()
    emitter.emit('pause')
  }

  async function dispose() {
    await nodes.dispose()
    streamNode.dispose()
  }

  async function stop() {
    await dispose()
    emitter.emit('stop')

    const payload = await nodes.processor.result
    const result = AudioBlob.fromRaw(
      payload.blob,
      nodes.context.currentTime * 1000,
      [payload.peaks, 100, 0, 100]
    )

    emitter.emit('result', result)
    return result
  }

  async function restart() {
    await dispose()
    await start()
  }

  function switchStream(stream: MediaStream) {
    streamNode.use(stream)
  }

  async function switchDevice(data: Partial<MediaDeviceInfo>) {
    switchStream(await StreamUtil.get(data))
  }

  return Object.freeze({
    init,
    start,
    resume,
    pause,
    stop,
    dispose,
    restart,
    switchStream,
    switchDevice,
    events: emitter,

    get nodes() {
      return nodes
    },

    get state() {
      return getContextState(nodes.context)
    }
  })
}

export * from './processor/factories'
