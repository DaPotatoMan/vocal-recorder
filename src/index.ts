import type { AudioBlob } from './shared'
import mitt from 'mitt'
import { Encoder } from './encoder'
import { RuntimeError, StreamUtil } from './shared'

export * from './shared'

export class AudioRecorder {
  events = AudioRecorder.Events.use()

  #recorder?: MediaRecorder
  #encoder?: Encoder

  get #instance() {
    if (this.#recorder)
      return this.#recorder

    throw new RuntimeError('RECORDER_NOT_INIT')
  }

  get state() {
    const state = this.#recorder?.state ?? 'inactive'

    return {
      current: state,
      paused: state === 'paused',
      inactive: state === 'inactive',
      recording: state === 'recording'
    }
  }

  get stream() {
    return this.#recorder?.stream
  }

  async init(config: AudioRecorder.Config = {}) {
    // Dispose active recorder
    this.dispose()

    // Get stream
    const stream = config.stream instanceof MediaStream
      ? config.stream
      : await StreamUtil.get(config.stream)

    this.#recorder = this.#createRecorder(stream)
    this.#encoder = new Encoder(this.#recorder)

    await this.#encoder.ready
    this.events.emit('init')
  }

  #createRecorder(stream: MediaStream) {
    const recorder = new MediaRecorder(stream)
    const eventKeys = ['start', 'stop', 'pause', 'resume', 'error'] as const

    // Delegate recorder events
    eventKeys.forEach(key =>
      recorder.addEventListener(key, (event) => {
        // Dispose stream when recorder stops
        if (key === 'stop') {
          StreamUtil.dispose(stream)
        }

        this.events.emit(key, event instanceof ErrorEvent ? event.error : event)
      })
    )

    return recorder
  }

  dispose() {
    if (this.#recorder) {
      if (this.#recorder.state !== 'inactive')
        this.#recorder.stop()

      this.#recorder = undefined
    }

    if (this.#encoder) {
      this.#encoder.dispose()
      this.#encoder = undefined
    }
  }

  start() {
    this.#encoder?.start()
  }

  pause() { this.#instance.pause() }
  resume() { this.#instance.resume() }

  async stop() {
    const result = await this.#encoder?.stop()

    if (!result)
      throw new RuntimeError('RECORDER_NO_RESULT')

    this.events.emit('result', result)
    return result
  }
}

export namespace AudioRecorder {
  export interface Config {
    stream?: MediaStream | MediaTrackConstraints
  }

  export namespace Events {
    export type Map = object & {
      /** Recorder is ready to be used */
      init: void

      start: Event
      stop: Event
      pause: Event
      resume: Event

      error: Error | RuntimeError
      result: AudioBlob
    }

    export type Keys = keyof Map
    export const use = mitt<Map>
  }
}
