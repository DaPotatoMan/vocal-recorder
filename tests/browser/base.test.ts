import { sleep } from '@antfu/utils'
import { AudioRecorder } from '../../src'

describe('can record', () => {
  const recorder = new AudioRecorder()

  it('class inits properly', () => {
    expect(recorder.state).toMatchSnapshot()
  })

  it('can record', async () => {
    await recorder.init()
    expect(recorder.config).toMatchSnapshot()
    expect(recorder.state).toMatchSnapshot()

    await recorder.start()
    expect(recorder.state).toMatchSnapshot()

    await sleep(5000)

    const result = await recorder.stop()
    expect(recorder.state).toMatchSnapshot()

    // Validate result
    expect(result.codec).toMatchSnapshot()
    expect(result.peaks.length).not.toBe(0)
    expect(result.duration).greaterThanOrEqual(4500)
  }, 10_000)
})
