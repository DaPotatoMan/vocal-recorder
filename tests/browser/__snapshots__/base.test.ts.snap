// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`class: AudioRecorder > can record 1`] = `
{
  "current": "inactive",
  "inactive": true,
  "paused": false,
  "recording": false,
}
`;

exports[`class: AudioRecorder > can record 2`] = `
{
  "current": "recording",
  "inactive": false,
  "paused": false,
  "recording": true,
}
`;

exports[`class: AudioRecorder > can record 3`] = `
{
  "current": "paused",
  "inactive": false,
  "paused": true,
  "recording": false,
}
`;

exports[`class: AudioRecorder > can record 4`] = `
{
  "current": "recording",
  "inactive": false,
  "paused": false,
  "recording": true,
}
`;

exports[`class: AudioRecorder > can record 5`] = `
{
  "current": "inactive",
  "inactive": true,
  "paused": false,
  "recording": false,
}
`;
